app/src/utils/networkPreflight.ts

```diff
@@ -0,0 +1,80 @@
+import { type ExternalProvider } from 'streamr-client'
+import { toaster } from 'toasterhea'
+import SwitchNetworkModal from '../modals/SwitchNetworkModal'
+import getChainId from './web3/getChainId'
+import getWeb3 from './web3/getWeb3'
+import { Layer } from './Layer'
+
+interface Currency {
+    decimals: number
+    name: string
+    symbol: string
+}
+
+interface Network {
+    chainId: string
+    chainName: string
+    blockExplorerUrls?: string[]
+    nativeCurrency?: Currency
+    rpcUrls: string[]
+}
+
+type Chain = [chainId: number, network: Network]
+
+/**
+ * @TODO Chain lists and configs are in few different places around the app. My guess is,
+ * the it'd be best to use `@streamr/config` here.
+ */
+export const Matic: Chain = [
+    137,
+    {
+        chainId: '0x89',
+        blockExplorerUrls: ['https://polygonscan.com'],
+        chainName: 'Polygon Mainnet',
+        nativeCurrency: {
+            decimals: 18,
+            name: 'Matic',
+            symbol: 'MATIC',
+        },
+        rpcUrls: ['https://polygon-rpc.com'],
+    },
+]
+
+/**
+ *
+ * @param expectedChainId Expected network/chain ID, e.g. 137 for Polygon.
+ * @returns `true` if the utility changed the network, and `false` if it did nothing (we're already on the correct network).
+ */
+export default async function networkPreflight(expectedChainId: number) {
+    const provider = getWeb3().currentProvider as ExternalProvider
+
+    try {
+        if ((await getChainId()) === expectedChainId) {
+            return false
+        }
+
+        await toaster(SwitchNetworkModal, Layer.Modal).pop()
```

Comment:
This is the modal piece. `pop()` is async and settles when you call `onResolve` or `onReject` inside your toastable (`Modal` in our case).

If you dismiss the toast it'll fail (`onReject` called) and the whole preflight fails, too.

New modals can be triggered from anywhere.

Nice thing?

```tsx
interface Props {
    onResolve?: (name: string) => void
    onReject?: (reason?: any) => void
}

function NewToastable({ onReject, onResolve }: Props) {
    return (
        <>
            <button
                type="button"
                onClick={() => void onResolve?.('foo')}
            >
                Give it some foo!
            </button>
            <button
                type="button"
                onClick={() => void onReject?.()}
            >
                Nah
            </button>
        </>
    )
}

const bar = await toaster(NewToastable, "container id")
```

After you click the first button somehow, `bar` is equal `foo`, and most importantly, properly typed.